networkx/algorithms/euler.py

```diff
@@ -147,3 +158,203 @@ def eulerian_circuit(G, source=None):
+           G.in_degree(second) == G.out_degree(second) + 1:
+            start = second
+        elif G.out_degree(second) == G.in_degree(second) + 1 and \
+             G.in_degree(first) == G.out_degree(first) + 1:
+            start = first
+        else:
+            start = None
+
+    else:
+        start = check_odd[0]
+
+    return start
+
+
+def eulerian_path(G):
+    """Return a generator of the edges of an Eulerian path in ``G``.
+
+    Check if the graph ``G`` has an Eulerian path and return a
+    generator for the edges. If no path is available, raise an error.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    edges: generator
+        A generator that produces edges in the Eulerian path.
+
+    Raises
+    ------
+    NetworkXError: If the graph does not have an Eulerian path.
+
+    Examples
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')])
+    >>> len(list(nx.eulerian_path(G)))
+    5
+
+    >>> G = nx.DiGraph([(1, 2), (2, 3)])
+    >>> list(nx.eulerian_path(G))
+    [(1, 2), (2, 3)]
+
+    Notes
+    -----
+    Linear time algorithm, adapted from [1]_ and [3]_.
+    Information about Euler paths in [2]_.
+    Code for Eulerian circuit in [3]_.
+
+    Important: In [1], Euler path is in reverse order,
+    this implementation gives the path in correct order
+    as in [3]_ for Eulerian_circuit. The distinction for
+    directed graph is in using the in_degree neighbors, not the out
+    ones. for undirected, it is using itemgetter(1) for get_vertex,
+    which uses the correct vertex for this order. Also, every graph
+    has an even number of odd vertices by the Handshaking Theorem [4]_.
+
+    References
+    ----------
+    .. [1] http://www.graph-magics.com/articles/euler.php
+    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
+    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
+    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
+
+    """
+    if not is_semieulerian(G):
+        raise nx.NetworkXError("G does not have an Eulerian path.")
+
+    g = G.__class__(G)  # copy graph structure (not attributes)
+    is_directed = g.is_directed()
+
+    if is_directed:
+        degree = g.in_degree
+        edges = g.in_edges
+        get_vertex = itemgetter(0)
+    else:
+        degree = g.degree
+        edges = g.edges
+        get_vertex = itemgetter(1)
+
+    # Begin algorithm:
+    start = _find_path_start(g)
+    if not start:
+        raise nx.NetworkXError("G has no Eulerian path.")
+
+    vertex_stack = deque([start])
+    last_vertex = None
+
+    while vertex_stack:
+
+        current_vertex = vertex_stack[-1]
+        # if no neighbors:
+        if degree(current_vertex) == 0:
+            # Special case, we cannot add a None vertex to the path.
```

Comment:
If this is a special case, maybe you can do the first iteration of this loop outside of the loop, so that it doesn't get checked every time through the loop.